#!/usr/bin/env python

PACKAGE="realsense2_camera"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

#             Name                                           Type       Level Description                     Default      Min          Max
gen.add("sr300_color_backlight_compensation",                int_t,     1,    "Backlight Compensation",       0,           0,           4)
gen.add("sr300_color_brightness",                            int_t,     2,    "Brightness",                   0,           -64,         64)
gen.add("sr300_color_contrast",                              int_t,     3,    "Contrast",                     50,          0,           100)
gen.add("sr300_color_gain",                                  int_t,     4,    "Gain",                         64,          0,           128)
gen.add("sr300_color_gamma",                                 int_t,     5,    "Gamma",                        300,         100,         500)
gen.add("sr300_color_hue",                                   int_t,     6,    "Hue",                          0,           -180,        180)
gen.add("sr300_color_saturation",                            int_t,     7,    "Saturation",                   64,          0,           100)
gen.add("sr300_color_sharpness",                             int_t,     8,    "Sharpness",                    50,          0,           100)

# Must be set only if color_enable_auto_white_balance is disabled
gen.add("sr300_color_white_balance",                         int_t,     9,    "White Balance",                4600,        2800,        6500)

gen.add("sr300_color_enable_auto_white_balance",             bool_t,    10,   "Enable Auto White Balance",    True)

# Must be set only if color_enable_auto_exposure is disabled
gen.add("sr300_color_exposure",                              int_t,     11,   "Exposure",                     156,         39,          10000)

gen.add("sr300_color_enable_auto_exposure",                  bool_t,    12,   "Enable Auto Exposure",         True)

preset_enum = gen.enum([gen.const("ShortRange",             int_t,  0,  "Short Range"),
                        gen.const("LongRange",              int_t,  1,  "Long Range"),
                        gen.const("BackgroundSegmentation", int_t,  2,  "Background Segmentation"),
                        gen.const("GestureRecognition",     int_t,  3,  "Gesture Recognition"),
                        gen.const("ObjectScanning",         int_t,  4,  "Object Scanning"),
                        gen.const("FaceAnalytics",          int_t,  5,  "Face Analytics"),
                        gen.const("FaceLogin",              int_t,  6,  "Face Login"),
                        gen.const("GrCursor",               int_t,  7,  "GR Cursor"),
                        gen.const("Default",                int_t,  8,  "Default"),
                        gen.const("MidRange",               int_t,  9,  "Mid Range"),
                        gen.const("IrOnly",                 int_t,  10, "IR Only")], "SR300 Visual Presets")

gen.add("sr300_depth_visual_preset",                         int_t,     13,   "SR300 Visual Presets",         8,           0,           10, edit_method=preset_enum)
gen.add("sr300_depth_laser_power",                           int_t,     14,   "Laser Power",                  16,          0,           16)
gen.add("sr300_depth_accuracy",                              int_t,     15,   "Accuracy",                     1,           1,           3)
gen.add("sr300_depth_motion_range",                          int_t,     16,   "Motion Range",                 9,           0,           220)
gen.add("sr300_depth_filter_option",                         int_t,     17,   "Filter Option",                5,           0,           7)
gen.add("sr300_depth_confidence_threshold",                  int_t,     18,   "Confidence Threshold",         3,           0,           15)
gen.add("sr300_depth_frames_queue_size",                     int_t,     19,   "Frames Queue Size",            16,          0,           32)
gen.add("sr300_depth_units",                                 double_t,  20,   "Depth Units",                  0.000124987, 0.000124987, 0.000124987)


# TODO
# Following are the Auto range options
#gen.add("sr300_auto_range_enable_motion_versus_range", int_t,  13,   "Enable Motion Versus Range",   1,         0,      2)
#gen.add("sr300_auto_range_enable_laser",               int_t,  14,   "Enable Laser",                 1,         0,      1)

# Must be set only if sr300_auto_range_enable_motion_versus_range is set to 1
#gen.add("sr300_auto_range_min_motion_versus_range",    int_t,  15,   "Min Motion Versus Range",      180,       -32767, 32767)
#gen.add("sr300_auto_range_max_motion_versus_range",    int_t,  16,   "Max Motion Versus Range",      605,       -32767, 32767)
#gen.add("sr300_auto_range_start_motion_versus_range",  int_t,  17,   "Start Motion Versus Range",    303,       -32767, 32767)

# Must be set only if sr300_auto_range_enable_laser is enabled
#gen.add("sr300_auto_range_min_laser",                  int_t,  18,   "Min Laser",                    2,         -32767, 32767)
#gen.add("sr300_auto_range_max_laser",                  int_t,  19,   "Max Laser",                    16,        -32767, 32767)
#gen.add("sr300_auto_range_start_laser",                int_t,  20,   "Start Laser",                  -1,        -32767, 32767)

#gen.add("sr300_auto_range_upper_threshold",            int_t,  21,   "Upper Threshold",              1250,      0,      65535)
#gen.add("sr300_auto_range_lower_threshold",            int_t,  22,   "Lower Threshold",              650,       0,      65535)

exit(gen.generate(PACKAGE, "realsense2_camera", "sr300_params"))

